clients/drcachesim/tools/reuse_distance.h

```diff
@@ -169,286 +169,416 @@ class reuse_distance_t : public analysis_tool_t {
     addr_t tag;
 
-    // We have a one-layer skip list for more efficient depth computation.
-    // We inline the fields in every node for simplicity and to reduce allocs.
-    struct line_ref_t *prev_skip; // the prev line_ref in the skip list
-    struct line_ref_t *next_skip; // the next line_ref in the skip list
-    int64_t depth;                // only valid for skip list nodes; -1 for others
-
-    line_ref_t(addr_t val)
-        : prev(NULL)
-        , next(NULL)
+    line_ref_node_t(addr_t val)
+        : left(nullptr)
+        , right(nullptr)
+        , parent(nullptr)
+        , size(1)
         , total_refs(1)
         , distant_refs(0)
         , tag(val)
-        , prev_skip(NULL)
-        , next_skip(NULL)
-        , depth(-1)
     {
     }
 };
 
-// We use a doubly linked list to keep track of the cache line reuse distance.
-// The head of the list is the most recently accessed cache line.
-// The earlier a cache line was accessed last time, the deeper that cache line
-// is in the list.
+// We use a splay to keep track of the cache line reuse distance.
+// The lefmost node of the splay tree is the most recently accessed cache line.
+// The earlier the cache line was last accessed, the more to the right this cache line
+// is in the splay tree.
 // If a cache line is accessed, its time stamp is set as current, and it is
-// added/moved to the front of the list.  The cache line reference
-// reuse distance is the cache line position in the list before moving.
+// added/moved to the left of the splay tree.  The cache line reference
+// reuse distance is the cache line position in the splay tree before moving.
 // We also keep a pointer (gate) pointing to the earliest cache
 // line referenced within the threshold.  Thus, we can quickly check
 // whether a cache line is recently accessed by comparing the time
 // stamp of the referenced cache line and the gate cache line.
-//
-// We have a second doubly-linked list, a one-layer skip list, for
-// more efficient computation of the depth.  Each node in the skip
-// list stores its depth from the front.
-struct line_ref_list_t {
-    line_ref_t *head_;       // the most recently accessed cache line
-    line_ref_t *gate_;       // the earliest cache line refs within the threshold
-    line_ref_t *tail_;       // the least recently accessed cache line
-    uint64_t cur_time_;      // current time stamp
-    uint64_t unique_lines_;  // the total number of unique cache lines accessed
-    uint64_t threshold_;     // the reuse distance threshold
-    uint64_t skip_distance_; // distance between skip list nodes
-    bool verify_skip_;       // check results using brute-force walks
-
-    line_ref_list_t(uint64_t reuse_threshold_, uint64_t skip_dist, bool verify)
-        : head_(NULL)
-        , gate_(NULL)
-        , tail_(NULL)
+// The splay tree is a binary search tree that using splay operatio for balancing, it
```

Comment:
> The splay tree is a binary search tree that using splay operatio for balancing, it
> allow to delete and insert an element at any position in O(log n) amortizated time.

Reword for grammar:

The splay tree is a binary search tree that uses the splay operation for balancing, which
allows deleting and inserting an element at any position in O(log n) amortized time.